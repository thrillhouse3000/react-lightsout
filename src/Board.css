.Board-table {
  border-spacing: 0.5rem;
}

.Board-winText {
  color: white;
  text-shadow: rgba(252, 255, 169, 0.795) 0px 0px 29px;
}