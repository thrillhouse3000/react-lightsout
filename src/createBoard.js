/** create a board nrows high/ncols wide, each cell randomly lit or unlit */
// TODO: create array-of-arrays of true/false values
const createBoard = {
    create(rows, cols) {
        let initialBoard = [];
        let values = [true, false]
        for (let i = 0; i < rows; i++) {
            let row = []
            for (let j = 0; j < cols; j++ ) {
                row.push({on: values[Math.floor(Math.random() * values.length)], coord: [i, j]})
            }
            initialBoard.push(row)
        }
        return initialBoard;
    }
}

export default createBoard;