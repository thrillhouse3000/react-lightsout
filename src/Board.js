import { useState } from "react";
import Cell from "./Cell";
import "./Board.css";
import createBoard from './createBoard'

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/


function Board({ nrows=3, ncols=3, chanceLightStartsOn=0.5 }) {
  const [board, setBoard] = useState(createBoard.create(nrows, ncols));

  // TODO: check the board in state to determine whether the player has won.\
  function hasWon() {
    let isWinner = false;
    board.every(row => (
      row.every(
        cell => cell.on === false) ? isWinner = true : isWinner = false
      )
    )
    return isWinner 
  }

  // TODO: Make a (deep) copy of the oldBoard
  // TODO: in the copy, flip this cell and the cells around it
  // TODO: return the copy
  const flipCellsAround = (coord) => {
    const [y,x] = coord
    function flip(cell) {
      if (cell.on === true) {
        return false
      } else {
        return true
      }
    }
    setBoard(
      board.map(row => (
        row.map(cell => (
          (cell.coord[0] === y && cell.coord[1] === x) ||
          (cell.coord[0] === y+1 && cell.coord[1] === x) || 
          (cell.coord[0] === y && cell.coord[1] === x+1) ||
          (cell.coord[0] === y-1 && cell.coord[1] === x) ||
          (cell.coord[0] === y && cell.coord[1] === x-1) ?
          {...cell, on: flip(cell) } :
          cell
        )
      )))
    )  
  }

  // TODO
  // if the game is won, just show a winning msg & render nothing else
  // make table board
  let winner = hasWon()
 
  return (
    <div className='Board'>
      {winner === true? (
        <h1 className="Board-winText">You Win!</h1>
      ) : (
        <table className='Board-table'>
          <tbody>
            {board.map((row, idx) => (
              <tr key={idx}>
                {row.map(cell => (
                  <Cell key={cell.coord} flipCellsAroundMe={flipCellsAround} isLit={cell.on} coord={cell.coord} />
                ))}
              </tr>
            ))}
          </tbody>
        </table>
      )}
    </div>
  )
}

export default Board;
