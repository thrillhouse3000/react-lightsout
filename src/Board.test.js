import React from 'react'
import { render, fireEvent } from '@testing-library/react';
import Board from './Board'
import createBoard from './createBoard'

beforeEach(() => {
    jest.spyOn(createBoard, 'create').mockImplementation(() => [
                    [{on: true, coord:[0, 0]}, {on: false, coord:[0, 1]}, {on: true, coord:[0, 2]}], 
                    [{on: false, coord:[1, 0]}, {on: true, coord:[1, 1]}, {on: false, coord:[1, 2]}], 
                    [{on: true, coord:[2, 0]}, {on: false, coord:[2, 1]}, {on: true, coord:[2, 2]}]
                ]);
});

afterEach(() => {
    createBoard.create.mockRestore();
})

it('renders without crashing', () => {
    render(<Board />)
})

it('should match snapshot', () => {
    const {asFragment} = render(<Board />)
    expect(asFragment()).toMatchSnapshot()
})

it('should switch the correct cells on click', () => {
    const {queryByTestId} = render(<Board/>);
    const center = queryByTestId("1,1")
    const up = queryByTestId("0,1")
    const down = queryByTestId("2,1")
    const left = queryByTestId("1,0")
    const right = queryByTestId("1,2")

    expect(center).toHaveClass('Cell', 'Cell-lit')
    expect(up).toHaveClass('Cell')
    expect(down).toHaveClass('Cell')
    expect(left).toHaveClass('Cell')
    expect(right).toHaveClass('Cell')

    fireEvent.click(center)
    expect(center).toHaveClass('Cell')
    expect(up).toHaveClass('Cell', 'Cell-lit')
    expect(down).toHaveClass('Cell', 'Cell-lit')
    expect(left).toHaveClass('Cell', 'Cell-lit')
    expect(right).toHaveClass('Cell', 'Cell-lit')
})