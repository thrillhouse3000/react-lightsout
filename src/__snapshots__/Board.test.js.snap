// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="Board"
  >
    <table
      class="Board-table"
    >
      <tbody>
        <tr>
          <td
            class="Cell Cell-lit"
            data-testid="0,0"
          />
          <td
            class="Cell "
            data-testid="0,1"
          />
          <td
            class="Cell Cell-lit"
            data-testid="0,2"
          />
        </tr>
        <tr>
          <td
            class="Cell "
            data-testid="1,0"
          />
          <td
            class="Cell Cell-lit"
            data-testid="1,1"
          />
          <td
            class="Cell "
            data-testid="1,2"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
            data-testid="2,0"
          />
          <td
            class="Cell "
            data-testid="2,1"
          />
          <td
            class="Cell Cell-lit"
            data-testid="2,2"
          />
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
